### PR TITLE
WIP: Refactor TypeScript plugins

### PR DESCRIPTION
**Pull request by dotansimha**
Related: https://github.com/dotansimha/graphql-code-generator/issues/1350 

Planned in this PR:
- [x] Move all common code from `flow` to a common package.
- [x] Make sure `flow` plugin works.
- [x] Add the latest changes of `flow` plugin.
- [x] Create `typescript` package with the same features as `flow` package.
- [x] Write tests for `typescript` package.
 
- [x] Move all common code from `flow-documents` to a common package.
- [x] Make sure `flow-documents` plugin works.
- [x] Add the latest changes of `flow-documents` plugin.
- [x] Create `typescript-documents` package with the same features as `flow-documents` package.
- [x] Write tests for `typescript-documents` package.
- [x] Move all common code from `flow-resolvers` to a common package.
- [x] Make sure `flow-resolvers` plugin works.
- [x] Add the latest changes of `flow-resolvers` plugin.
- [x] Create `typescript-resolvers` package with the same features as `flow-resolvers` package.
- [x] Write tests for `typescript-resolvers` package.
- [x] Add missing features from previous `typescript` package.
- [x] Add missing features from previous `typescript-documents` package.
- [x] Add missing features from previous `typescript-resolvers` package.
- [x] Add introspection types @kamilkisiela  https://github.com/dotansimha/graphql-code-generator/pull/1406
- [x] Implement `namingConvention` again @kamilkisiela  https://github.com/dotansimha/graphql-code-generator/pull/1407
- [x] Fix issue with prefix @kamilkisiela https://github.com/dotansimha/graphql-code-generator/pull/1407
- [x] Apply this: https://github.com/dotansimha/graphql-code-generator/pull/1357 @ardatan 
- [x] Refactor `react-apollo` @dotansimha 
- [x] Refactor `apollo-angular` @dotansimha 
- [x] Apply fix to Flow plugins: https://github.com/dotansimha/graphql-code-generator/pull/1413 @kamilkisiela 
- [x] Refactor `typescript-mongo` @dotansimha  
- [x] Refactor `stencil-apollo` @ardatan
- [x] Remove Handlebars compiler package from all plugins, and add support for tree shaking. @ardatan 
- [x] Make sure to update packages in `gql-gen init`  @kamilkisiela 
- [x] Rename `XXX-documents` to `XXX-operations`
- [x] Move all flatten utils from `core` to `plugin-helpers` and clean `core` package. @kamilkisiela 
- [x] Deprecate `buildTemplateContext` and `flattenTypes`. @kamilkisiela 
- [x] Repo cleanup
- [x] Update README @kamilkisiela  
- [x] Update docs @kamilkisiela  
- [x] Write docs about writing plugins with visitor @kamilkisiela 
- [x] Update examples
- [x] Change packages names @dotansimha

### CODE REVIEW

**Comment by kamilkisiela** [comment]
- [ ] Make sure to update packages in `gql-gen init`

File: packages/plugins/typescript-documents/tests/ts-documents.spec.ts [hunk 1]
```diff
@@ -845,5 +845,55 @@ describe('TypeScript Documents Plugin', async () => {
...
+      const content = await plugin(
+        testSchema,
+        [{ filePath: '', content: query }],
+        { typesPrefix: 'PREFIX_' },
+        {
+          outputFile: 'graphql.ts'
+        }
+      );
+
+      // Kamil: is `PREFIX_Prefix_*` correct?
```
Comment (kamilkisiela) [inline packages/plugins/typescript-documents/tests/ts-documents.spec.ts]: @dotansimha ?

Comment (dotansimha) [reply packages/plugins/typescript-documents/tests/ts-documents.spec.ts]: Nope, it should be just `PREFIX_*`. It seems like we apply the prefix twice, one without `convertName` and one with it.

Comment (kamilkisiela) [reply packages/plugins/typescript-documents/tests/ts-documents.spec.ts]: I'm on it

**Comment by kamilkisiela** [comment]
> add support for tree shaking
Actually, `ng-packagr` works nice and produces multiple bundles to multiple targets with CJS and ESM. But it might use `tsc` wrapped by angular/compiler-cli or even don't support TS 3.3+. But I think it worth to check it. It also publishes everything from a directory so the distributed package uses root directory to store files. It also takes care of `package.json` and its fields.

@dotansimha @ardatan

**Pull request merged**